src/Libraries/Microsoft.Extensions.AI.OpenAI/DataContentAsyncEnumerableStream.cs

```diff
@@ -0,0 +1,206 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Shared.Diagnostics;
+
+namespace Microsoft.Extensions.AI;
+
+/// <summary>
+/// Utility class to stream <see cref="IAsyncEnumerable{T}"/> data content as a <see cref="Stream"/>.
+/// </summary>
+#if !NET8_0_OR_GREATER
+internal sealed class DataContentAsyncEnumerableStream : Stream, IAsyncDisposable
+#else
+internal sealed class DataContentAsyncEnumerableStream : Stream
+#endif
+{
+    private readonly IAsyncEnumerator<DataContent> _enumerator;
+    private bool _isCompleted;
+    private ReadOnlyMemory<byte>? _remainingData;
+    private int _remainingDataOffset;
+    private long _position;
+    private DataContent? _firstDataContent;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="DataContentAsyncEnumerableStream"/> class/>.
+    /// </summary>
+    /// <param name="dataAsyncEnumerable">The async enumerable to stream.</param>
+    /// <param name="firstDataContent">The first chunk of data to reconsider when reading.</param>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
+    /// <remarks>
+    /// <paramref name="firstDataContent"/> needs to be considered back in the stream if <paramref name="dataAsyncEnumerable"/> was iterated before creating the stream.
+    /// This can happen to check if the first enumerable item contains data or is just a reference only content.
+    /// </remarks>
+    internal DataContentAsyncEnumerableStream(IAsyncEnumerable<DataContent> dataAsyncEnumerable, DataContent? firstDataContent = null, CancellationToken cancellationToken = default)
+    {
+        _enumerator = Throw.IfNull(dataAsyncEnumerable).GetAsyncEnumerator(cancellationToken);
+        _remainingData = Memory<byte>.Empty;
+        _remainingDataOffset = 0;
+        _position = 0;
+        _firstDataContent = firstDataContent;
+    }
+
+    /// <inheritdoc/>
+    public override bool CanRead => true;
+
+    /// <inheritdoc/>
+    public override bool CanSeek => false;
+
+    /// <inheritdoc/>
+    public override bool CanWrite => false;
+
+    /// <inheritdoc/>
+    public override long Length => throw new NotSupportedException();
+
+    /// <inheritdoc/>
+    public override async Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
```

Comment:
Is this override needed? It looks like what the base implementation will end up doing.